src/Npgsql/PostgresDatabaseInfo.cs

```diff
@@ -201,161 +203,139 @@ internal async Task<List<PostgresType>> LoadBackendTypes(NpgsqlConnector conn, N
-                    new(GenerateLoadTypesQuery(SupportsRangeTypes, conn.Settings.LoadTableComposites)),
-                    new(GenerateLoadCompositeTypesQuery(conn.Settings.LoadTableComposites)),
-                }
-            };
+            var batchQuery = new StringBuilder("SELECT version();")
+                .Append(GenerateLoadTypesQuery(SupportsRangeTypes, conn.Settings.LoadTableComposites))
+                .Append(GenerateLoadCompositeTypesQuery(conn.Settings.LoadTableComposites));
 
             if (SupportsEnumTypes)
-                batch.BatchCommands.Add(new(GenerateLoadEnumFieldsQuery(HasEnumSortOrder)));
-
-            batch.Timeout = commandTimeout;
-            batch.AllResultTypesAreUnknown = true;
+                batchQuery.Append(GenerateLoadEnumFieldsQuery(HasEnumSortOrder));
 
+            conn.WriteBuffer.Timeout = TimeSpan.FromSeconds(commandTimeout);
             timeout.CheckAndApply(conn);
-            var reader = async ? await batch.ExecuteReaderAsync() : batch.ExecuteReader();
-            try
-            {
-                var byOID = new Dictionary<uint, PostgresType>();
 
-                // First the PostgreSQL version
-                if (async)
-                {
-                    await reader.ReadAsync();
-                    LongVersion = reader.GetString(0);
-                    await reader.NextResultAsync();
-                }
-                else
+            var q = batchQuery.ToString();
+            Debug.WriteLine($"Type loading query:{Environment.NewLine}{q}");
+            await conn.WriteQuery(q, async);
+            await conn.Flush(async);
+            var byOID = new Dictionary<uint, PostgresType>();
+            var buf = conn.ReadBuffer;
+
+            // First read the PostgreSQL version
+            var rowDescription = Expect<RowDescriptionMessage>(await conn.ReadMessage(async), conn);
+            Debug.Assert(rowDescription.Count == 1, "The 'version query' should return exactly one column.");
+
+            // We read the message in non-sequential mode which buffers the whole message.
+            // There is no need to ensure data within the message boundaries
+            Expect<DataRowMessage>(await conn.ReadMessage(async), conn);
+            buf.Skip(2);
+            LongVersion = buf.ReadString(buf.ReadInt32());
+            Expect<CommandCompleteMessage>(await conn.ReadMessage(async), conn);
+
+            // Then load the types
+            rowDescription = Expect<RowDescriptionMessage>(await conn.ReadMessage(async), conn);
+            Debug.Assert(rowDescription.Count == 6, "The 'type loading query' should return exactly 6 columns.");
+            IBackendMessage msg;
+            while (true)
+            {
+                msg = await conn.ReadMessage(async);
+                if (msg is not DataRowMessage)
+                    break;
+
+                buf.Skip(2);
+                var nspname = buf.ReadString(buf.ReadInt32());
```

Comment:
Can do a small local method to combine `buf.ReadString(buf.ReadInt32())`, which is repeated a lot here.